src/OptimizedPathMappingRepository.php

```diff
@@ -0,0 +1,365 @@
+        }
+
+        // Read children before attaching the resource to this repository
+        $children = $resource->listChildren();
+
+        $resource->attachTo($this, $path);
+
+        // Add the resource before adding its children, so that the array
+        // stays sorted
+        $this->store->set($path, $resource);
+
+        $basePath = '/' === $path ? $path : $path.'/';
+
+        foreach ($children as $name => $child) {
+            $this->addResource($basePath.$name, $child);
+        }
+    }
+
+
+    /**
+     * @param FilesystemResource $resource
+     */
+    private function removeResource(FilesystemResource $resource)
+    {
+        $path = $resource->getPath();
+
+        // Ignore non-existing resources
+        if (! $this->store->exists($path)) {
+            return;
+        }
+
+        // Recursively register directory contents
+        foreach ($this->iteratorToCollection($this->getChildIterator($resource)) as $child) {
+            $this->removeResource($child);
+        }
+
+        $this->store->remove($path);
+
+        // Detach from locator
+        $resource->detach($this);
+    }
+
+    /**
+     * Returns an iterator for the children of a resource.
+     *
+     * @param FilesystemResource $resource The resource.
+     *
+     * @return RegexFilterIterator|Resource[] The iterator.
+     */
+    private function getChildIterator(FilesystemResource $resource)
+    {
+        $staticPrefix = rtrim($resource->getPath(), '/').'/';
+        $regExp = '~^'.preg_quote($staticPrefix, '~').'[^/]+$~';
+
+        return new RegexFilterIterator(
+            $regExp,
+            $staticPrefix,
+            new \ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Returns an iterator for a glob.
+     *
+     * @param string $glob The glob.
+     *
+     * @return GlobFilterIterator|Resource[] The iterator.
+     */
+    private function getGlobIterator($glob)
+    {
+        return new GlobFilterIterator(
+            $glob,
+            new \ArrayIterator($this->store->keys())
+        );
+    }
+
+    /**
+     * Recursively creates a directory for a path.
+     *
+     * @param string $path A directory path.
+     * @return DirectoryResource The created resource
+     */
+    private function ensureDirectoryExists($path)
+    {
+        if (! $this->store->exists($path)) {
+            // Recursively initialize parent directories
+            if ($path !== '/') {
+                $this->ensureDirectoryExists(Path::getDirectory($path));
+            }
+
+            $resource = new GenericFilesystemResource($path);
+            $resource->attachTo($this, $path);
+
+            $this->store->set($path, $resource);
+
+            return;
+        }
+    }
+
+    /**
+     * Transform an iterator of paths into a collection of resources
+     *
+     * @param \Iterator $iterator
+     * @return FilesystemResourceCollection
+     */
+    private function iteratorToCollection(\Iterator $iterator)
```

Comment:
As `getGlobalIterator()` and `getChildIterator()` return paths, this method transform these path into a collection of resources.